### PR TITLE
Parametrize `Deserialize{Value,Row}` with two lifetimes: `'frame` and `'metadata` separately

### PR DESCRIPTION
**Pull request by wprzytula**
Refs: #462

## Tl;dr
A second lifetime parameter is introduced to entities comprising the bottom part of the new deserialization framework: `'metadata`.

## Deserialization traits - recap

`DeserializeValue` requires two types of data in order to perform deserialization:
1) a reference to the CQL frame (a `FrameSlice`),
2) the type of the column being deserialized, being part of the result metadata.

Similarly, `DeserializeRow` requires two types of data in order to perform deserialization:
1) a reference to the CQL frame (a `FrameSlice`),
2) a slice of specifications of all columns in the row, being part of the result metadata.

### Observation

When deserializing owned types, both the frame and the metadata can have any lifetime and it's not important. When deserializing borrowed types, however, they borrow from the frame, so their lifetime must necessarily be bound by the lifetime of the frame. Metadata is only needed for the deserialization, so its lifetime does not abstractly bound the deserialized value.

### Problem

Up to this PR, `Deserialize{Value, Row}` were only parametrized by one lifetime: `'frame`, which bounded both the frame slice and the metadata. This was done that way due to an assumption that both the metadata and the frame (shared using `Bytes`) will be owned by the same entity. However, the new idea of deserializing result metadata to a borrowed form (to save allocations) makes result metadata's lifetime shorter than the frame's lifetime.

### Implemented solution

Not to unnecessarily shorten the deserialized values' lifetime, a separate lifetime parameter is introduced for result metadata: `'metadata`.

The PR is large, but the changes are mostly mechanical, by adding
the second lifetime parameter.

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~~[ ] I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- [ ] I have adjusted the documentation in `./docs/source/`. <-- _not yet, will be done at the end of the deserialization refactor_
- ~~[ ] I added appropriate `Fixes:` annotations to PR description.~~

### CODE REVIEW

**Comment by wprzytula** [comment]
> `cargo semver-checks` found no API-breaking changes in this PR! 🎉🥳 Checked commit: [4406db5](https://github.com/scylladb/scylla-rust-driver/commit/4406db55710c161a35188629bfe348e38e9eafa6)

???

This looks like a crude bug... Adding a new lifetime parameter to a trait necessarily breaks the API, in the simplest possible way: syntactically.

File: scylla-cql/src/types/deserialize/value.rs [hunk 1]
```diff
@@ -216,34 +216,36 @@ impl_fixed_numeric_type!(f64, Double);
...
-    Ok(num_bigint_03::BigInt::from_signed_bytes_be(val))
-});
+impl_emptiable_strict_type!(
+    num_bigint_03::BigInt,
+    Varint,
+    |typ: &'metadata ColumnType<'metadata>, v: Option<FrameSlice<'frame>>| {
+        let val = ensure_not_null_slice::<Self>(typ, v)?;
+        Ok(num_bigint_03::BigInt::from_signed_bytes_be(val))
+    }
+);
```
Comment (muzarski) [inline scylla-cql/src/types/deserialize/value.rs:233]: The old odd formatting got fixed. Nice!

File: scylla-cql/src/types/deserialize/value.rs [hunk 2]
```diff
@@ -436,7 +438,9 @@ fn get_days_since_epoch_from_date_column<T>(
 }
 
 #[cfg(feature = "chrono-04")]
-impl_emptiable_strict_type!(chrono_04::NaiveDate, Date, |typ: &'frame ColumnType,
+impl_emptiable_strict_type!(chrono_04::NaiveDate, Date, |typ: &'metadata ColumnType<
+    'metadata,
+>,
```
Comment (muzarski) [inline scylla-cql/src/types/deserialize/value.rs:443]: The formatter going crazy again

**Review by muzarski — APPROVED** [review]
LGTM. Great explanation of why we need to introduce second lifetime parameter, so the review went smoothly

**Comment by Lorak-mmk** [comment]
> Refs: #462
> ## Tl;dr
> 
> A second lifetime parameter is introduced to entities comprising the bottom part of the new deserialization framework: `'metadata`.
> ## Deserialization traits - recap
> 
> `DeserializeValue` requires two types of data in order to perform deserialization:
> 
>     1. a reference to the CQL frame (a `FrameSlice`),
> 
>     2. the type of the column being deserialized, being part of the result metadata.
> 
> 
> Similarly, `DeserializeRow` requires two types of data in order to perform deserialization:
> 
>     1. a reference to the CQL frame (a `FrameSlice`),
> 
>     2. a slice of specifications of all columns in the row, being part of the result metadata.
> 
> 
> ### Observation
> 
> When deserializing owned types, both the frame and the metadata can have any lifetime and it's not important. When deserializing borrowed types, however, they borrow from the frame, so their lifetime must necessarily be bound by the lifetime of the frame. Metadata is only needed for the deserialization, so its lifetime does not abstractly bound the deserialized value.

If that were the case then there would be no need to introduce second lifetime to those traits, you could just use anonymous lifetime / have a lifetime-generic method:
```
    fn deserialize(
        typ: &ColumnType,
        v: Option<FrameSlice<'frame>>,
    ) -> Result<Self, DeserializationError>;
```

The reason that seems not possible it precisely that deserialized values are bound by metadata lifetime, for example a `ListlikeIterator`.

This sentence confused me so much that I had to go and try to do this change (and learn why it is not possible) in order to understand why this PR introduces the second lifetime. 
Please fix the sentence.

> ### Problem
> 
> Up to this PR, `Deserialize{Value, Row}` were only parametrized by one lifetime: `'frame`, which bounded both the frame slice and the metadata. This was done that way due to an assumption that both the metadata and the frame (shared using `Bytes`) will be owned by the same entity. However, the new idea of deserializing result metadata to a borrowed form (to save allocations) makes result metadata's lifetime shorter than the frame's lifetime.
> ### Implemented solution
> 
> Not to unnecessarily shorten the deserialized values' lifetime, a separate lifetime parameter is introduced for result metadata: `'metadata`.
> 
> The PR is large, but the changes are mostly mechanical, by adding the second lifetime parameter.

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]
Suggestion: such change should have new tests, right? Those should showcase usage with 2 different lifetimes.

**Comment by wprzytula** [comment]
v1.1: addressed comments.

**Review by Lorak-mmk — APPROVED** [review]
